### PR TITLE
pallet-xcm: Deprecate `execute` and `send` in favor of `execute_blob` and `send_blob`

### PR DESCRIPTION
**Pull request by franciscoaguirre**
`execute` and `send` try to decode the xcm in the parameters before reaching the filter line.
The new extrinsics decode only after the filter line.
These should be used instead of the old ones.

## TODO
- [x] Tests
- [x] Generate weights
- [x] Deprecation issue -> https://github.com/paritytech/polkadot-sdk/issues/3771
- [x] PRDoc
- [x] Handle error in pallet-contracts

This would make writing XCMs in PJS Apps more difficult, but here's the fix for that: https://github.com/polkadot-js/apps/pull/10350.
Already deployed! https://polkadot.js.org/apps/#/utilities/xcm

Supersedes https://github.com/paritytech/polkadot-sdk/pull/1798/

### CODE REVIEW

**Comment by franciscoaguirre** [comment]
@pgherveou Could you take a look at `ExecuteController` and `SendController` and make sure the changes make sense? I don't want to break anything on the contracts side

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5575054

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5575073

File: polkadot/xcm/xcm-builder/src/controller.rs
Comment (pgherveou) [inline polkadot/xcm/xcm-builder/src/controller.rs:95]: this should probably not belong to the trait, just let it be an implementation detail of whatever implement this

File: polkadot/xcm/xcm-builder/src/controller.rs
Comment (pgherveou) [inline polkadot/xcm/xcm-builder/src/controller.rs:64]: maybe just use a `#[deprecated*note = "..."]` attribute instead of these WARNING?

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (pgherveou) [inline polkadot/xcm/pallet-xcm/src/lib.rs:385]:
```suggestion
					log::error!(target: "xcm::send_blob", "Unable to decode XCM, error: {:?}", error);
```

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (pgherveou) [inline polkadot/xcm/pallet-xcm/src/lib.rs:1507]: Where is MaxXcmEncodedSize, is this supposed to come from from the pallet config?

**Comment by pgherveou** [comment]
> @pgherveou Could you take a look at `ExecuteController` and `SendController` and make sure the changes make sense? I don't want to break anything on the contracts side

this should work, I am not too sure where your new MaxXcmEncodedSize is defined though. If that helps we can also not add the legacy function to the traits since at that point the contract implementation is still unstable and ok to change.

**Comment by franciscoaguirre** [comment]
I'll put only the new function in the traits then and ping you later to check everything would be fine with contracts.

File: polkadot/xcm/xcm-builder/src/controller.rs
Comment (franciscoaguirre) [reply polkadot/xcm/xcm-builder/src/controller.rs:64]: I put it in the doc comment since those get added to the metadata, and UIs won't see the rust deprecation warnings but they will see the metadata.
I could add two, one for the UIs and one for internal use

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (franciscoaguirre) [reply polkadot/xcm/pallet-xcm/src/lib.rs:1507]: Yes, pallet config. Right now I had it hardcoded, but leaving it as a config shouldn't be bad

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5582785

**Comment by acatangiu** [comment]
New PJS utility to allow same UX for these new extrinsics is already available, thanks to @serban300 👍 
https://polkadot.js.org/apps/#/utilities/xcm

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5586825

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5590163

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5590485

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5590483

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5591539

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5591558

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (pgherveou) [inline polkadot/xcm/pallet-xcm/src/lib.rs:1509]: you don't do this here (and in send_blob) below?
```
			let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin)?;
```

**Comment by pgherveou** [comment]
Can you tell a bit more about this
> execute and send try to decode the xcm in the parameters before reaching the filter line.

The new interface is a bit awkward, the caller needs to create an XCM, then encode it, so that it's later decoded and processed inside the pallet. 

Couldn't we just get away with an extra encoded_size() check on the passed message to see if it fits the constraint?

**Comment by franciscoaguirre** [comment]
You only encode and then decode because you're writing the code in rust, but normal users would already have to encode their XCM

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5598478

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5600638

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (franciscoaguirre) [reply polkadot/xcm/pallet-xcm/src/lib.rs:1509]: That's already done in `ExecuteController::execute_blob` (and `SendController::send_blob`)

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (franciscoaguirre) [reply polkadot/xcm/pallet-xcm/src/lib.rs]: I updated the value based on a test and put a link to it in the doc comment

**Comment by franciscoaguirre** [comment]
bot help

**Comment by franciscoaguirre** [comment]
bot bench cumulus-assets --runtime asset-hub-westend --pallet pallet_xcm
bot bench cumulus-assets --runtime asset-hub-rococo --pallet pallet_xcm
bot bench cumulus-collectives --pallet pallet_xcm
bot bench cumulus-coretime --runtime coretime-rococo --pallet pallet_xcm
bot bench cumulus-coretime --runtime coretime-westend --pallet pallet_xcm
bot bench cumulus-bridge-hubs --runtime bridge-hub-rococo --pallet pallet_xcm
bot bench cumulus-bridge-hubs --runtime bridge-hub-westend --pallet pallet_xcm
bot bench cumulus-people --runtime people-rococo --pallet pallet_xcm
bot bench cumulus-people --runtime people-westend --pallet pallet_xcm
bot bench polkadot-pallet --runtime rococo --pallet pallet_xcm
bot bench polkadot-pallet --runtime westend --pallet pallet_xcm

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5606854

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5606872

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5607866

**Comment by franciscoaguirre** [comment]
bot clean

**Comment by franciscoaguirre** [comment]
bot bench cumulus-assets --runtime asset-hub-westend --pallet pallet_xcm
bot bench cumulus-assets --runtime asset-hub-rococo --pallet pallet_xcm
bot bench cumulus-collectives --pallet pallet_xcm
bot bench cumulus-coretime --runtime coretime-rococo --pallet pallet_xcm
bot bench cumulus-coretime --runtime coretime-westend --pallet pallet_xcm
bot bench cumulus-bridge-hubs --runtime bridge-hub-rococo --pallet pallet_xcm
bot bench cumulus-bridge-hubs --runtime bridge-hub-westend --pallet pallet_xcm
bot bench cumulus-people --runtime people-rococo --pallet pallet_xcm
bot bench cumulus-people --runtime people-westend --pallet pallet_xcm
bot bench polkadot-pallet --runtime rococo --pallet pallet_xcm
bot bench polkadot-pallet --runtime westend --pallet pallet_xcm

File: substrate/frame/contracts/src/wasm/runtime.rs
Comment (franciscoaguirre) [inline substrate/frame/contracts/src/wasm/runtime.rs:2170]: @pgherveou Any idea?

**Comment by franciscoaguirre** [comment]
bot bench cumulus-bridge-hubs --runtime bridge-hub-westend --pallet pallet_xcm

File: substrate/frame/contracts/src/wasm/runtime.rs
Comment (pgherveou) [reply substrate/frame/contracts/src/wasm/runtime.rs:2170]: Will look into it shortly feel free to leave a todo if that's unblock you

Comment (pgherveou) [reply substrate/frame/contracts/src/wasm/runtime.rs:2170]: We should adjust the interface here, we should read a BoundedVec<u8, MaxXcmEncodedSize> instead of VersionedXcm<()>

I can give it a shot and push a commit to your PR, early next week

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (pgherveou) [inline polkadot/xcm/pallet-xcm/src/lib.rs:1497]: missing doc string here and below on send_blob

File: polkadot/xcm/xcm-builder/src/controller.rs
Comment (pgherveou) [inline polkadot/xcm/xcm-builder/src/controller.rs:101]:
```suggestion
	/// - `msg`: the encoded XCM to be sent, should be decodable as a [`VersionedXcm`]
```

File: polkadot/xcm/xcm-builder/src/controller.rs
Comment (pgherveou) [inline polkadot/xcm/xcm-builder/src/controller.rs:69]:
```suggestion
	/// - `msg`: the encoded XCM to be executed, should be decodable as a [`VersionedXcm`]
```

**Comment by franciscoaguirre** [comment]
Had to force push because I added unrelated files on one of the last commits

**Comment by franciscoaguirre** [comment]
@acatangiu @pgherveou It's ready for final review 😄

File: cumulus/parachains/integration-tests/emulated/common/src/impls.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/common/src/impls.rs:376]: nit:
```suggestion
							xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/send.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/send.rs:81]:
```suggestion
			xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/send.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/send.rs:165]:
```suggestion
			xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/swap.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/swap.rs:376]:
```suggestion
			xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/send_xcm.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/send_xcm.rs:45]:
```suggestion
			xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs:102]:
```suggestion
			remote_xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs:176]:
```suggestion
			create_agent_xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs:185]:
```suggestion
			create_channel_xcm.encode().try_into().unwrap(),
```

File: cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/send_xcm.rs
Comment (acatangiu) [inline cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/send_xcm.rs:45]:
```suggestion
			xcm.encode().try_into().unwrap(),
```

File: polkadot/runtime/rococo/src/impls.rs
Comment (acatangiu) [inline polkadot/runtime/rococo/src/impls.rs:177]: let's not `expect` in runtime code, even if we know it should work, easy enough to just return an error

File: polkadot/runtime/westend/src/impls.rs
Comment (acatangiu) [inline polkadot/runtime/westend/src/impls.rs:177]: same here, return an error rather than panicking

File: polkadot/xcm/pallet-xcm/src/lib.rs
Comment (acatangiu) [inline polkadot/xcm/pallet-xcm/src/lib.rs:1077]:
```suggestion
		/// WARNING: DEPRECATED. `execute` will be removed after June 2024. Use `execute_blob`
		/// instead.
		#[allow(deprecated)]
		#[deprecated(note = "`execute` will be removed after June 2024. Use `execute_blob` instead.")]
```

File: polkadot/xcm/pallet-xcm/src/tests/mod.rs
Comment (acatangiu) [inline polkadot/xcm/pallet-xcm/src/tests/mod.rs:312]:
```suggestion
		let encoded_versioned_message = versioned_message.encode().try_into().unwrap();
```

File: polkadot/xcm/pallet-xcm/src/tests/mod.rs
Comment (acatangiu) [inline polkadot/xcm/pallet-xcm/src/tests/mod.rs:360]:
```suggestion
				VersionedXcm::from(message.clone()).encode().try_into().unwrap(),
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5645075

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
@pgherveou @bkontur this is ready, waiting for final review

**Review by pgherveou — APPROVED** [review]

**Pull request merged**
